### PR TITLE
Branches - Monick

### PR DESCRIPTION
**Pull request by mfunkemomo**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I did a custom model method for spotlight and top_ten categories.
Describe how you approached testing that model method. What edge cases did you come up with? | I tested that it printed 10 works and made sure that it was sorted with the works with the most votes at the top and the least at the bottom. 
What are `session` and `flash`? What is the difference between them? | Session is a way for devs to communicate with the browser as it will keep track if a user is logged in or not. Session will last as long as the cookies are on and the user hasn't logged out. Flash will only last for one response-request cycle and is just a way for devs to communicate with users their status and give them a response. 
What was one thing that you gained more clarity on through this assignment? | How rails comes together as a whole and how there are ways to help dry code but there are many files.
What is the Heroku URL of your deployed application? | https://momos-media-ranker.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? |  While it's intimidating at first, I thought it was a good overall project and not too stressful. I would have liked more time to dry up code and apply newly learned things from last week but ran out of time to apply those things.

### CODE REVIEW

File: app/views/layouts/application.html.erb [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+            <%= link_to "Add a new work", new_work_path, method: :get %>
+          </li>
+          <li class="nav-item app-header__nav_item">
+            <%= link_to "View all users", users_path, method: :get %>
+          </li>
+        </ul>
+        <ul class="nav app-header__user-nav-container">
+          <% if session[:user_id] %>
+            <li class="nav-item app-header__nav_item">
+              <p><%= link_to "Logged in as #{find_user.username}", user_path(session[:user_id]), method: :get %></p>
```
Comment (jmaddox19) [inline app/views/layouts/application.html.erb:37]: The logic here is what could go in the determine_user method, assigning it to an instance variable. 
`@current_user = user_path(session[:user_id])`

Comment (jmaddox19) [reply app/views/layouts/application.html.erb:37]: That way it can be accessible from any controller or view.

**Comment by jmaddox19** [comment]
# Media Ranker
## What We're Looking For



### Manual Testing

Workflow | yes / no
--- | ---
Deployed to Heroku | Yes!
Optional: Look and feel is similar to the original (consider styling and layout) | Yes! Wow! Remarkably so!
**Before logging in** |
On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | Yes, other than movies but not a problem :)
Can go into a work's show page | X
Verify unable to vote on a work, and get a flash message | As a guest, I cannot upvote but there is no flash message
Can edit this work successfully, and get a flash message | Yes, but no flash message.
Can go to "View all media" page and see three lists of works, sorted by vote | Yes
Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | Yes!
Can create a new work successfully. Note the URL for this work's show page | Yes, but there's a weird additional input field under title that I think might be an accident
Can delete this work successfully | Yes
Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | Yes
Verify that the "View all users" page lists no users (except what is produced by db seeding) | Yes
**Log in** | Yes
Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | Yes
Your username is listed in "View all users" page | Yes
Verify that number of votes determines the Media Spotlight | Yes
Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | Yes
Voting on the same work twice produces an error/flash message, and there is no extra vote | No flash message but the user's 2nd vote isn't recorded
**Log out** |
Logging out showed a flash message and changed the UI | Yes!
Logging in as a new user creates a new user | Yes!
Logging in as an already existing user has a specific flash message | Yes

### Targeted Code Review

Area | yes (X) / no
--- | ---
git commits were small and atomic, with useful messages | X
Routes file uses `resources` for works | X
Routes file shows intention in limiting routes for voting, log-in functionality, and users | I'm not sure any of the votes routes are actually used.
The homepage view, all media view, and new works view use semantic HTML | X
The model for media (likely named `work.rb`) `has_many` votes | X
The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | X
Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | Yes, but it's empty, so not actually doing anything. (See comment in PR)
Some controller, likely the `WorksController`, has a controller filter for finding a work | X
The `WorksController` uses strong params | Yes, but work_params is in ApplicationController. It would be more appropriate in WorksController.
The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | Yes!

#### Targed Test Review

Area | yes / no
--- | ---
There are valid fixtures files used for users, votes, and works | X
User model has tests with sections on validations (valid and invalid) and relationships (has votes) | No tests for relationships
Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | No tests for relationships
Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | No tests for relationships
Work model has tests with a section on all business logic methods in the model, including their edge cases | Yes

## Overall Feedback
Great job! Please do read through all of the notes in this rubric and my comments on your code in GitHub to see all the little pieces of feedback I had to give. You made a full website with user sign-in and stuff! So awesome!
